xarray/core/missing.py

```diff
@@ -6,13 +6,44 @@
 import pandas as pd
 
 from . import utils
-from .common import _contains_datetime_like_objects
+from .common import _contains_datetime_like_objects, ones_like
 from .computation import apply_ufunc
 from .duck_array_ops import dask_array_type
 from .utils import OrderedSet, is_scalar
 from .variable import Variable, broadcast_variables
 
 
+def _get_nan_block_lengths(obj, dim, index):
+    """
+    Return an object where each NaN element in 'obj' is replaced by the
+    length of the gap the element is in.
+    """
+
+    # algorithm from https://stackoverflow.com/questions/53060003/how-to-get-the-maximum-time-of-gap-in-xarray-dataset/53075828#53075828
+    arange = ones_like(obj) * index + (index[1] - index[0])
+    cumulative_nans = arange.where(obj.notnull()).ffill(dim=dim).fillna(0)
+
+    num_nans = arange - cumulative_nans
+
+    block_lengths_at_peaks = num_nans.where(
+        num_nans.diff(dim=dim, label="lower") < 0
+    ).reindex({dim: obj[dim]})
+
+    # nans at the end
+    maybe_nans_at_end = (
+        block_lengths_at_peaks.isel({dim: -1})
+        .where(obj.isel({dim: -1}).notnull(), num_nans.isel({dim: -1}))
+        .expand_dims(dim)
+        .reindex_like(obj)
+    )
+
+    block_lengths_at_peaks = block_lengths_at_peaks.fillna(maybe_nans_at_end)
+
+    nan_block_lengths = block_lengths_at_peaks.bfill(dim).where(obj.isnull()).fillna(0)
+
+    return nan_block_lengths
```

Comment:
Here is a version of the code that works with monotonic irregularly sampled data:
```python
valid = obj.notnulls()
varange = arange.where(valid)
cumulative_nans = varange.ffill(dim=dim).fillna(index[0])

nan_block_lengths = cumulative_nans_.diff(dim=dim, label='upper').reindex({dim: obj[dim]}).where(valid).bfill(dim=dim).where(~valid, 0).fillna(index[-1]-varange.max())
```
It can be probably optimised.
The exact size of the gaps at edges cannot be properly computed, so the lower bound is computed with available coordinates.